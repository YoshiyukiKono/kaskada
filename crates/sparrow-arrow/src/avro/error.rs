use arrow::datatypes::DataType;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "unimplemented conversion from Arrow type {_0:?} to Avro")]
    Unimplemented(DataType),
}

impl error_stack::Context for Error {}
