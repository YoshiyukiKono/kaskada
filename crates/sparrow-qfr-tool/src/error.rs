use sparrow_qfr::kaskada::sparrow::v1alpha::flight_record::Record;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "internal error")]
    Internal,
    #[display(fmt = "unsupported flight record: {_0:?}")]
    UnsupportedRecord(Option<Record>),
    #[display(fmt = "undefined activity ID")]
    UndefinedActivityId,
    #[display(fmt = "undefined metric ID")]
    UndefinedMetricId,
    #[display(fmt = "missing metric value")]
    MissingMetricValue,
}

impl error_stack::Context for Error {}
