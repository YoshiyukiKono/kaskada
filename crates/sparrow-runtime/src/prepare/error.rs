use std::path::PathBuf;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "unable to open '{path:?}'")]
    OpenFile { path: PathBuf },
    #[display(fmt = "unable to create '{path:?}'")]
    CreateFile { path: PathBuf },
    #[display(fmt = "internal error")]
    Internal,
    #[display(fmt = "failed to create Parquet file reader")]
    CreateParquetReader,
    #[display(fmt = "failed to create CSV file reader")]
    CreateCsvReader,
    #[display(fmt = "reading batch")]
    ReadingBatch,
    #[display(fmt = "slicing batch")]
    SlicingBatch,
    #[display(fmt = "non-nullable column '{field}' had {null_count} nulls")]
    NullInNonNullableColumn { field: String, null_count: usize },
    #[display(fmt = "preparing column")]
    PreparingColumn,
    #[display(fmt = "sorting batch")]
    SortingBatch,
    #[display(fmt = "determine metadata")]
    DetermineMetadata,
    #[display(fmt = "failed to write Parquet file")]
    WriteParquetData,
    #[display(fmt = "failed to write metadata file")]
    WriteMetadata,
    #[display(fmt = "prepare request missing '{_0}'")]
    MissingField(&'static str),
    #[display(
        fmt = "invalid prepare request: slice plan table should match table config, but was '{slice_plan}' and '{table_config}'"
    )]
    IncorrectSlicePlan {
        slice_plan: String,
        table_config: String,
    },
    #[display(fmt = "unsupported file path {_0:?}")]
    UnsupportedOutputPath(sparrow_api::kaskada::v1alpha::file_path::Path),
    #[display(fmt = "invalid input path")]
    InvalidInputPath,
    #[display(fmt = "failed to read input")]
    ReadInput,
    #[display(fmt = "failed to upload result")]
    UploadResult,
}

impl error_stack::Context for Error {}

impl sparrow_core::ErrorCode for Error {
    fn error_code(&self) -> tonic::Code {
        match self {
            Self::MissingField(_) | Self::IncorrectSlicePlan { .. } | Self::InvalidInputPath => {
                tonic::Code::InvalidArgument
            }
            Self::UnsupportedOutputPath(_) => tonic::Code::Unimplemented,
            _ => tonic::Code::Internal,
        }
    }
}
