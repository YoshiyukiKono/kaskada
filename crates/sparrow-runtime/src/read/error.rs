use std::sync::Arc;

use chrono::NaiveDateTime;

use crate::data_manager::DataHandle;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "failed to create input stream")]
    CreateStream,
    #[display(fmt = "failed to read next batch")]
    ReadNextBatch,
    #[allow(unused)]
    #[display(fmt = "unsupported: {_0}")]
    Unsupported(&'static str),
    #[display(fmt = "internal error: ")]
    Internal,
    #[display(
        fmt = "internal error: min next time ({min_next_time}) must be <= next lower bound {lower_bound}"
    )]
    MinNextGreaterThanNextLowerBound {
        min_next_time: i64,
        lower_bound: i64,
    },
    #[display(
        fmt = "internal error: max event time ({max_event_time}) must be > next upper bound {upper_bound}"
    )]
    MaxEventTimeLessThanNextUpperBound {
        max_event_time: i64,
        upper_bound: i64,
    },
    #[display(fmt = "failed to select necessary prepared files")]
    SelectPreparedFiles,
    #[display(fmt = "failed to queue download of necessary prepared files")]
    QueueFileDownloads,
    #[display(
        fmt = "unexpected file '{file:?}' in table '{table_name}' with data partially before the snapshot time {snapshot_time:?}"
    )]
    PartialOverlap {
        file: Arc<DataHandle>,
        table_name: String,
        snapshot_time: Option<NaiveDateTime>,
    },
    #[display(fmt = "failed to skip to minimum event")]
    SkippingToMinEvent,
    #[display(fmt = "failed to load table schema")]
    LoadTableSchema,
    #[display(fmt = "failed to determine projected schema")]
    DetermineProjectedSchema,
}

impl error_stack::Context for Error {}
