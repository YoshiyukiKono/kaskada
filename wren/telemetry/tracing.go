package telemetry

import (
	"context"
	"reflect"

	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
	"github.com/rs/zerolog/log"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	resource "go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/genproto/googleapis/rpc/errdetails"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"

	v1alpha "github.com/kaskada-ai/kaskada/gen/proto/go/kaskada/kaskada/v1alpha"
)

// TracingProvider injects the current request trace-id into logs and error responses
type TracingProvider interface {
	UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (response interface{}, err error)
	StreamInterceptor(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error
}

type tracingProvider struct{}

// NewTracingProvider creates a new tracing provider designed for GRPC Interceptors (currently only Unary)
// returns the provider and a shutdown function
func NewTracingProvider(ctx context.Context, otelEndpont string) (TracingProvider, func()) {
	var tracerProvider *sdktrace.TracerProvider
	if otelEndpont == "" {
		// when not exporting traces, still provide a generic traceProvider
		// to generate requestIDs, which are needed for logging to work correctly
		tracerProvider = sdktrace.NewTracerProvider()
	} else {
		// hook up exporter, don't block on connection to collector
		traceExporter, err := otlptracegrpc.New(ctx,
			otlptracegrpc.WithInsecure(),
			otlptracegrpc.WithEndpoint(otelEndpont),
		)
		if err != nil {
			log.Ctx(ctx).Error().Err(err).Msg("failed to create trace exporter")
			return nil, func() {}
		}

		// autodetect resource attributes
		resources, err := resource.New(ctx,
			// ignore the docs, `resource.WithBuiltinDetectors` is deprecated
			// https://opentelemetry.io/docs/go/exporting_data/#resources - not updated as of 2021/08/26
			// https://github.com/open-telemetry/opentelemetry-go/issues/2026
			resource.WithHost(),
			resource.WithOSDescription(),
			resource.WithOSType(),
			resource.WithProcess(),
			// Or specify resource attributes directly
			// resource.WithAttributes(attribute.String("foo", "bar")),
		)
		if err != nil {
			log.Ctx(ctx).Error().Err(err).Msg("failed to create resources for trace")
		}

		// Register the trace exporter with a TracerProvider, using a batch
		// span processor to aggregate spans before export.
		bsp := sdktrace.NewBatchSpanProcessor(traceExporter)
		tracerProvider = sdktrace.NewTracerProvider(
			sdktrace.WithResource(resources),
			sdktrace.WithSampler(sdktrace.AlwaysSample()),
			sdktrace.WithSpanProcessor(bsp),
		)
	}
	otel.SetTracerProvider(tracerProvider)

	// set global propagator to tracecontext (the default is no-op).
	propagator := propagation.NewCompositeTextMapPropagator(propagation.Baggage{}, propagation.TraceContext{})
	otel.SetTextMapPropagator(propagator)

	shutdownFunction := func() {
		// Shutdown will flush any remaining spans and shut down the exporter.
		err := tracerProvider.Shutdown(ctx)
		if err != nil {
			log.Ctx(ctx).Error().Err(err).Msg("issue shutting down trace provider")
		} else {
			log.Ctx(ctx).Info().Msg("cleanly shut down trace provider")
		}
	}

	return &tracingProvider{}, shutdownFunction
}

// UnaryInterceptor intercepts unary calls and adds the trace-id to the context logger as request_id
func (tp *tracingProvider) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (response interface{}, err error) {
	traceID := getTraceID(ctx)
	ctx = addTraceIDToContextAndLogs(ctx, traceID)

	// Perform the request
	res, err := handler(ctx, req)

	return addRequestDetailsToResponse(res, traceID), addRequestDetailsToError(err, traceID)
}

// StreamInterceptor intercepts stream calls and adds the trace-id to the context logger as request_id
func (tp *tracingProvider) StreamInterceptor(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
	t := newTracingServerStream(ss)
	err := handler(srv, t)
	return addRequestDetailsToError(err, t.traceID)
}

func getTraceID(ctx context.Context) string {
	return trace.SpanFromContext(ctx).SpanContext().TraceID().String()
}

func addTraceIDToContextAndLogs(ctx context.Context, traceID string) context.Context {
	// add trace-id as request_id to tracing interceptor for request & payload logs
	ctx = logging.InjectFields(ctx, logging.Fields{"request_id", traceID})

	// add trace-id as request_id to context logger for all other logs
	contextLogger := log.With().Str("request_id", traceID).Logger()
	return contextLogger.WithContext(ctx)
}

func addRequestDetailsToResponse(res interface{}, traceID string) interface{} {
	// attempt to add the trace-id to the RequestDetails of the response
	if res != nil {
		resVal := reflect.ValueOf(res)
		resVal = reflect.Indirect(resVal)
		if resVal.IsValid() && resVal.Kind() == reflect.Struct {
			requestDetails := resVal.FieldByName("RequestDetails")

			// add a RequestDetails object to the repsonse if one doesn't already exist
			if requestDetails.IsValid() && requestDetails.IsNil() && requestDetails.CanSet() {

				switch {
				case reflect.TypeOf(&v1alpha.RequestDetails{}).AssignableTo(requestDetails.Type()):
					requestDetails.Set(reflect.ValueOf(&v1alpha.RequestDetails{}))
				default:
					log.Error().Msg("unknown type for addRequestDetailsToResponse")
				}
			}

			// add the RequestId to the RequestDetails object
			requestDetails = reflect.Indirect(requestDetails)
			if requestDetails.Kind() == reflect.Struct {
				requestID := requestDetails.FieldByName("RequestId")
				if requestID.IsValid() && requestID.CanSet() {
					requestID.Set(reflect.ValueOf(traceID))
				}
			}
		}
	}
	return res
}

func addRequestDetailsToError(err error, traceID string) error {
	if err != nil {
		s, ok := status.FromError(err)
		if !ok {
			// From: https://pkg.go.dev/google.golang.org/grpc@v1.39.0#UnaryHandler

			// UnaryHandler defines the handler invoked by UnaryServerInterceptor to complete the normal execution
			// of a unary RPC. If a UnaryHandler returns an error, it should be produced by the status package, or
			// else gRPC will use codes.Unknown as the status code and err.Error() as the status message of the RPC.
			s = status.New(codes.Unknown, err.Error())
		}
		s, err = s.WithDetails(&errdetails.RequestInfo{
			RequestId: traceID,
		})
		if err == nil {
			return s.Err()
		}
	}
	return err
}

type tracingServerStream struct {
	ctx     context.Context
	ss      grpc.ServerStream
	traceID string
}

func newTracingServerStream(ss grpc.ServerStream) tracingServerStream {
	traceID := getTraceID(ss.Context())
	return tracingServerStream{
		ctx:     addTraceIDToContextAndLogs(ss.Context(), traceID),
		ss:      ss,
		traceID: traceID,
	}
}

func (t tracingServerStream) SetHeader(m metadata.MD) error {
	return t.ss.SetHeader(m)
}

func (t tracingServerStream) SendHeader(m metadata.MD) error {
	return t.ss.SendHeader(m)
}

func (t tracingServerStream) SetTrailer(m metadata.MD) {
	t.ss.SetTrailer(m)
}

func (t tracingServerStream) Context() context.Context {
	return t.ctx
}

func (t tracingServerStream) SendMsg(m interface{}) error {
	m = addRequestDetailsToResponse(m, t.traceID)
	err := t.ss.SendMsg(m)
	return addRequestDetailsToError(err, t.traceID)
}

func (t tracingServerStream) RecvMsg(m interface{}) error {
	return t.ss.RecvMsg(m)
}
