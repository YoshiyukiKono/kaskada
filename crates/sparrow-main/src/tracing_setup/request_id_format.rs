use std::io;

use opentelemetry::trace::TraceContextExt;
use serde::ser::{SerializeMap, Serializer as _};
use tracing::{Event, Subscriber};
use tracing_opentelemetry::OtelData;
use tracing_serde::{AsSerde, SerdeMapVisitor};
use tracing_subscriber::fmt::format::Writer;
use tracing_subscriber::fmt::{FmtContext, FormatEvent, FormatFields};
use tracing_subscriber::registry::{LookupSpan, SpanRef};

/// Custom format for JSON logging events a `request_id`.
pub struct RequestIdFormat;

impl<S, N> FormatEvent<S, N> for RequestIdFormat
where
    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
    N: for<'writer> FormatFields<'writer> + 'static,
{
    fn format_event(
        &self,
        ctx: &FmtContext<'_, S, N>,
        mut writer: Writer<'_>,
        event: &Event<'_>,
    ) -> std::fmt::Result
    where
        S: Subscriber + for<'a> LookupSpan<'a>,
    {
        let meta = event.metadata();

        let mut visit = || {
            let mut serializer = serde_json::Serializer::new(WriteAdaptor::new(&mut writer));

            let mut serializer = serializer.serialize_map(None)?;
            serializer.serialize_entry("timestamp", &chrono::Utc::now())?;
            serializer.serialize_entry("level", &meta.level().as_serde())?;
            serializer.serialize_entry("line", &meta.line())?;
            serializer.serialize_entry("module", &meta.module_path())?;
            serializer.serialize_entry("target", meta.target())?;

            if let Some(trace_info) = ctx.lookup_current().as_ref().and_then(lookup_trace_info) {
                serializer
                    .serialize_entry("request_id", &hex::encode(trace_info.trace_id.to_bytes()))?;
                serializer
                    .serialize_entry("span_id", &hex::encode(trace_info.span_id.to_bytes()))?;
            }

            let mut visitor = SerdeMapVisitor::new(serializer);
            event.record(&mut visitor);
            visitor.finish()
        };

        visit().map_err(|_| std::fmt::Error)?;
        writeln!(writer)
    }
}

struct WriteAdaptor<'a> {
    fmt_write: &'a mut dyn std::fmt::Write,
}

impl<'a> WriteAdaptor<'a> {
    fn new(fmt_write: &'a mut dyn std::fmt::Write) -> Self {
        Self { fmt_write }
    }
}

impl<'a> io::Write for WriteAdaptor<'a> {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        let s =
            std::str::from_utf8(buf).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;

        self.fmt_write
            .write_str(s)
            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;

        Ok(s.as_bytes().len())
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

pub struct TraceInfo {
    pub trace_id: opentelemetry::trace::TraceId,
    pub span_id: opentelemetry::trace::SpanId,
}

fn lookup_trace_info<S>(span_ref: &SpanRef<'_, S>) -> Option<TraceInfo>
where
    S: Subscriber + for<'a> LookupSpan<'a>,
{
    span_ref
        .extensions()
        .get::<OtelData>()
        .map(|otel| TraceInfo {
            trace_id: otel.parent_cx.span().span_context().trace_id(),
            span_id: otel
                .builder
                .span_id
                .unwrap_or(opentelemetry::trace::SpanId::INVALID),
        })
}
