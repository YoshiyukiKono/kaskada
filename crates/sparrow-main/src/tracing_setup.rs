use std::time::Duration;

use opentelemetry::global;
use opentelemetry::propagation::Extractor;
use opentelemetry::sdk::propagation::TraceContextPropagator;
use opentelemetry::sdk::resource::ResourceDetector;
use opentelemetry::sdk::trace::Tracer;
use opentelemetry::sdk::Resource;
use opentelemetry_otlp::WithExportConfig;
use request_id_format::RequestIdFormat;
use tonic::codegen::http;
use tracing_opentelemetry::{OpenTelemetryLayer, OpenTelemetrySpanExt};
use tracing_subscriber::layer::SubscriberExt;
use tracing_subscriber::util::SubscriberInitExt;
use tracing_subscriber::EnvFilter;

use crate::BuildInfo;

mod request_id_format;

/// The options available for configuring tracing.
///
/// The Jaeger exporter should also support the environment variables documented
/// here: <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#jaeger-exporter>
///
/// The options `OTEL_EXPORTER_JAEGER_ENDPOINT`, `OTEL_EXPORTER_JAEGER_USER` and
/// `OTEL_EXPORTER_JAEGER_PASSWORD` are only supported if the `collector_client`
/// feature of the `opentelemetry-jaeger` crate are enabled.
// TODO: Consider replicating the supported options so they show up in --help
#[derive(clap::Args, Debug)]
pub struct TracingOptions {
    /// Log & tracing filter configuration.
    ///
    /// Defaults to `egg::=warn,sparrow_=trace,info` which omits most traces
    /// from the `egg` crate (because it's very log-heavy), and `trace` from
    /// any of the sparrow crates (`sparrow_runtime`, `sparrow_main`, etc.),
    /// and `info` from other crates.
    #[arg(
        long,
        default_value = "egg::=warn,sparrow_=trace,info",
        env = "SPARROW_LOG_FILTER"
    )]
    log_filters: String,
    /// Whether OpenTelemetry tracing should be enabled.
    #[arg(long, env = "SPARROW_OTEL_TRACING_ENABLED")]
    tracing_enabled: bool,
    /// Whether logs should be exported as JSON.
    ///
    /// When JSON logs are enabled we include the Trace ID in log
    /// messages and omit parent spans information.
    #[arg(long, env = "SPARROW_LOG_JSON")]
    log_json: bool,

    /// Whether to disable color output in logs.
    /// 
    /// Set to `1` to disable color output.
    #[arg(long, default_value = "0", env = "NO_COLOR")]
    log_no_color: i8,
}

impl TracingOptions {
    fn create_tracer<T>(&self) -> Option<OpenTelemetryLayer<T, Tracer>>
    where
        T: tracing::Subscriber + for<'span> tracing_subscriber::registry::LookupSpan<'span>,
    {
        if self.tracing_enabled {
            // Setup tracing reporting to open telemetry via OTLP.
            let resource = Resource::from_detectors(
                Duration::new(5, 0),
                vec![
                    Box::new(opentelemetry::sdk::resource::EnvResourceDetector::new()),
                    Box::new(opentelemetry::sdk::resource::OsResourceDetector),
                    Box::new(opentelemetry::sdk::resource::ProcessResourceDetector),
                    Box::new(opentelemetry::sdk::resource::SdkProvidedResourceDetector),
                    Box::new(SparrowBuildResourceDetector),
                ],
            );
            let tracer = opentelemetry_otlp::new_pipeline()
                .tracing()
                .with_trace_config(
                    opentelemetry::sdk::trace::Config::default().with_resource(resource),
                )
                .with_exporter(opentelemetry_otlp::new_exporter().tonic().with_env())
                .install_batch(opentelemetry::runtime::Tokio)
                .unwrap();

            // Configure propagation. This is only necessary for the gRPC serving, but
            // doesn't hurt other commands with nothing to propagate.
            global::set_text_map_propagator(TraceContextPropagator::new());

            Some(tracing_opentelemetry::layer().with_tracer(tracer))
        } else {
            None
        }
    }
}

/// Setup tracing to both Jaeger and local console.
///
/// # Panics
/// If any of the setup fails. This is called early on and is required to work
/// in order for logging / debugging / etc.
pub fn setup_tracing(config: &TracingOptions) {
    // Enable filtering based on `log_filters`.
    // Do this early so that we can exclude logs (and traces) from spammy components
    // such as egg.
    let registry = tracing_subscriber::registry()
        .with(EnvFilter::try_new(&config.log_filters).unwrap())
        .with(tracing_error::ErrorLayer::default());

    // Add logging and install the handler. We do this in one block
    // since the type of logs (JSON or not) affects the type of the
    // format layer and registry.
    if config.log_json {
        let fmt_layer = tracing_subscriber::fmt::layer().with_ansi(config.log_no_color != 1).event_format(RequestIdFormat);
        registry
            .with(fmt_layer)
            .with(config.create_tracer())
            .try_init()
            .unwrap();
    } else {
        let fmt_layer = tracing_subscriber::fmt::layer().with_ansi(config.log_no_color != 1).with_target(false);
        registry
            .with(fmt_layer)
            .with(config.create_tracer())
            .try_init()
            .unwrap();
    }
}

struct SparrowBuildResourceDetector;

impl ResourceDetector for SparrowBuildResourceDetector {
    fn detect(&self, _timeout: std::time::Duration) -> Resource {
        BuildInfo::default().as_resource()
    }
}

/// Create an initial span with the opentelemetry context
///
/// The intended use is as the `trace_fn` in the gRPC server. This ensures that
/// the request is executed in a properly connected span.
pub fn propagate_span(request: &http::Request<()>) -> tracing::Span {
    let header_map = HeaderMap(request.headers());
    let parent_cx = global::get_text_map_propagator(|prop| prop.extract(&header_map));
    let span = tracing::info_span!("Start of request");

    // Set the OTEL parent of the tracing span. This ties things into the
    // overall OTEL trace.
    span.set_parent(parent_cx);
    span
}

/// Wrapper forf extending HeaderMap.
struct HeaderMap<'a>(&'a http::header::HeaderMap);

impl<'a> Extractor for HeaderMap<'a> {
    fn get(&self, key: &str) -> Option<&str> {
        // Get a value for a key from the headers.
        // If the value can't be converted to &str, returns None
        self.0.get(key).and_then(|h| h.to_str().ok())
    }

    fn keys(&self) -> Vec<&str> {
        // Collect all the keys from the headers.
        self.0.keys().map(|k| k.as_str()).collect::<Vec<_>>()
    }
}
