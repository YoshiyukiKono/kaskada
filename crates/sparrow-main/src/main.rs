#![warn(
    rust_2018_idioms,
    nonstandard_style,
    future_incompatible,
    clippy::mod_module_files,
    clippy::print_stdout,
    clippy::print_stderr,
    clippy::undocumented_unsafe_blocks
)]

use clap::{command, Parser};
use error_stack::ResultExt;
use opentelemetry::global;
use sparrow_main::tracing_setup::{setup_tracing, TracingOptions};
use sparrow_main::{BatchCommand, PrepareCommand, ServeCommand};
use tracing::error;

#[cfg(not(target_os = "windows"))]
const NOTICE: &str = include_str!("../../../NOTICE");

#[cfg(target_os = "windows")]
const NOTICE: &str = include_str!("..\\..\\..\\NOTICE");

/// The top-level options for using Sparrow.
#[derive(clap::Parser, Debug)]
#[command(name = "sparrow", rename_all = "kebab-case", version)]
pub struct SparrowOptions {
    #[command(flatten)]
    tracing_options: TracingOptions,

    #[command(subcommand)]
    command: Command,
    /// Enables a custom panic handler which logs panics.
    ///
    /// This may be disabled to get the default backtrace and re-entrant
    /// panic handler.
    #[arg(long, env = "SPARROW_DISABLE_LOG_PANIC_HANDLER")]
    disable_log_panic_handler: bool,
}

#[derive(clap::Subcommand, Debug)]
#[allow(clippy::large_enum_variant)]
enum Command {
    /// Run the Sparrow gRPC service.
    Serve(ServeCommand),
    /// Run Sparrow in batch-mode on a specific script.
    Batch(BatchCommand),
    /// Prepare a file for use as part of a table.
    Prepare(PrepareCommand),
    /// License report and notice.
    License,
}

#[tokio::main]
async fn main() {
    let options = SparrowOptions::parse();
    setup_tracing(&options.tracing_options);

    if !options.disable_log_panic_handler {
        std::panic::set_hook(Box::new(logging_panic_hook));
    }

    let exit_code = if let Err(err) = main_body(options).await {
        error!("{:?}", err);
        1
    } else {
        0
    };

    // Exit with the given exit code.
    std::process::exit(exit_code);
}

#[derive(derive_more::Display, Debug)]
#[display(fmt = "error running command")]
pub struct Error;

impl error_stack::Context for Error {}

#[allow(clippy::print_stdout)]
async fn main_body(options: SparrowOptions) -> error_stack::Result<(), Error> {
    match options.command {
        Command::Serve(serve) => serve.execute().await.change_context(Error)?,
        Command::Batch(batch) => batch.execute().await.change_context(Error)?,
        Command::Prepare(prepare) => prepare.execute().change_context(Error)?,
        Command::License => {
            println!("{NOTICE}");
        }
    };

    Ok(())
}

fn logging_panic_hook(panic_info: &std::panic::PanicInfo<'_>) {
    let message = panic_info
        .payload()
        .downcast_ref::<&str>()
        .copied()
        .or_else(|| {
            panic_info
                .payload()
                .downcast_ref::<String>()
                .map(|s| s.as_ref())
        });

    // TODO: Look at the implementation of the default panic hook and pull some of
    // that over if needed. Specifically, logging of backtrace (if enabled)
    // and handling of "double panic" (eg., if flushing the tracer provider
    // panics).
    match (message, panic_info.location()) {
        (Some(message), Some(location)) => error!("Panic: {} at {}", message, location),
        (Some(message), None) => error!("Panic: {} at unknown location", message),
        (None, Some(location)) => error!("Panic: unknown at {}", location),
        (None, None) => error!("Panic: unknown at unknown location"),
    }

    // Flush any tracing spans before returning.
    // This is still likely to produce an abrupt termination.
    global::shutdown_tracer_provider();
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_main_help_stdout() {
        let mut cmd = assert_cmd::Command::cargo_bin("sparrow-main").unwrap();
        let assert = cmd.arg("--help").assert().success();
        let output = std::str::from_utf8(&assert.get_output().stdout).unwrap();
        insta::assert_snapshot!(output);
    }

    #[test]
    fn test_batch_help_stdout() {
        let mut cmd = assert_cmd::Command::cargo_bin("sparrow-main").unwrap();
        let assert = cmd.arg("batch").arg("--help").assert().success();
        let output = std::str::from_utf8(&assert.get_output().stdout).unwrap();
        insta::assert_snapshot!(output);
    }

    #[test]
    fn test_prepare_help_stdout() {
        let mut cmd = assert_cmd::Command::cargo_bin("sparrow-main").unwrap();
        let assert = cmd.arg("prepare").arg("--help").assert().success();
        let output = std::str::from_utf8(&assert.get_output().stdout).unwrap();
        insta::assert_snapshot!(output);
    }

    #[test]
    fn test_serve_help_stdout() {
        let mut cmd = assert_cmd::Command::cargo_bin("sparrow-main").unwrap();
        let assert = cmd.arg("serve").arg("--help").assert().success();
        let output = std::str::from_utf8(&assert.get_output().stdout).unwrap();
        insta::assert_snapshot!(output);
    }
}
