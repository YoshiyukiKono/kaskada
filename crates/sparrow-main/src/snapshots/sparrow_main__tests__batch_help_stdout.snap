---
source: crates/sparrow-main/src/main.rs
expression: output
---
Run Sparrow in batch-mode on a specific script

Usage: sparrow-main batch [OPTIONS] --schema <SCHEMA> --script <SCRIPT>

Options:
      --store-initial-dfg <STORE_INITIAL_DFG>
          Path to store a copy of the pre-optimization DFG to. Defaults to not storing anything

      --store-final-dfg <STORE_FINAL_DFG>
          Path to store a copy of the post-optimization DFG to. Defaults to not storing anything

      --store-plan-graph <STORE_PLAN_GRAPH>
          Path to store a copy of the query plan to as a `dot` file. Defaults to not storing a plan

      --store-plan-yaml <STORE_PLAN_YAML>
          Path to store a copy of the query plan to as a `yaml` file. Defaults to not storing a plan

      --simplifier-iteration-limit <SIMPLIFIER_ITERATION_LIMIT>
          Maximum number of iterations of the simplifier to run.
          
          NOTE: The environment variable won't affect the default values.
          
          [env: SPARROW_SIMPLIFIER_ITERATION_LIMIT=]
          [default: 30]

      --simplifier-node-limit <SIMPLIFIER_NODE_LIMIT>
          Maximum number of nodes during simplifier execution.
          
          NOTE: The environment variable won't affect the default values.
          
          [env: SPARROW_SIMPLIFIER_NODE_LIMIT=]
          [default: 10000]

      --simplifier-time-limit-seconds <SIMPLIFIER_TIME_LIMIT_SECONDS>
          Maximum amount of time to run the simplifier before ending early.
          
          NOTE: The environment variable won't affect the default values.
          
          [env: SPARROW_SIMPLIFIER_TIME_LIMIT_SECONDS=]
          [default: 5]

      --experimental
          Whether experimental behaviors should be enabled

      --preview-rows <PREVIEW_ROWS>
          Produces a preview of the data with at least this many rows.
          
          Default value (0) indicates all rows should be produced.
          
          [default: 0]

      --flight-record-path <FLIGHT_RECORD_PATH>
          Path to store the Query Flight Record to. Defaults to not storing anything

      --compile-only
          Only compile (and output the plan/etc. if requested)

      --schema <SCHEMA>
          File containing the schema definitions for the script

      --script <SCRIPT>
          Input file containing the script to run

      --output-dir <OUTPUT_DIR>
          Output directory to write the output to
          
          [default: .]

  -h, --help
          Print help (see a summary with '-h')

  -V, --version
          Print version

